Add CI jobs

cottsay opened the pull request:
TODO list for this feature:
- [x] Workspace bringup
- [x] Colcon build and test
- [x] Docker invalidation based on
  - Changes in the repos file
  - Changes in the rosdep index
  - Changes in the base OS' package versions
- [x] Output artifact for use as underlay for subsequent job
- [x] ccache
- [x] Jenkins parameters
- [x] Lots of other `ros_buildfarm_config` values (priority, etc)
- [x] Scheduling
- [x] Overlay job
- [x] Documentation
  - [x] Build file / Configuration - [configuration_options.rst](https://github.com/ros-infrastructure/ros_buildfarm/blob/ci-builds/doc/configuration_options.rst#specific-options-in-ci-build-files)
  - [x] Usage - [ci_jobs.rst](https://github.com/ros-infrastructure/ros_buildfarm/blob/ci-builds/doc/jobs/ci_jobs.rst)
- [x] Catkin support
- [x] Package selection
- [x] Non-Jenkins / Script support
- [x] Travis CI entry invoking the script version of the job
- [x] Solve the installed package enumeration issue in Colcon (remove the hacky workaround)
    This may not be as complicated as I thought - at least some of the issues I was trying to work around are actually package bugs that can be fixed.
- [x] Support for overlay job on the script version of the job (and subsequent Travis entry)
- [x] Solve or work around the locals vs globals issue in template expansion.
    This is needed to support referencing other variables in generator expressions.
- [x] Explore concurrency challenges on workers
    After discussing this, I think there will be some work to do here.

dirk-thomas commented:
As mentioned in https://github.com/ros-infrastructure/ros_buildfarm/pull/575#issuecomment-432431998 CI builds should work with `colcon` as well as `catkin_make_isolated` (depending on a configuration flag) - similar to the devel jobs. As a consequence several names (files, directories, functions) shouldn't contain `colcon` in. E.g. the workspace should be named just `ws` as it is for all the other job types. Also why is `colcon_build.py` necessary? What does it do beyond the existing `build_and_install` / `build_and_test` scripts? Are there any other parts where the existing parts for devel jobs can be reused instead of creating new ones for CI builds?

Please also create a `.rst` file for the new CI build file documenting the options. Looking at the code `build_ignore` seems to be a relative path. Using just the package name might be easier.

Please also create the "script" versions of the script in order to enable running the same build locally. This should then also be used in the Travis file to test do some CI builds ensuring it works.

dirk-thomas commented on ros_buildfarm/config/ci_build_file.py:
```diff
@@ -0,0 +1,71 @@
+            self.name
+        assert int(data['version']) in [1], \
+            ("Unable to handle '%s' format version '%d'." % \
+            (CIBuildFile._type, int(data['version']))
```
CI indicates a syntax error here. One closing parenthesis is missing.

cottsay commented:
Thanks @dirk-thomas - I updated the TODO list above per your feedback. Please adjust as necessary.

cottsay commented:
I believe this is ready for more eyes now. Some things I know aren't working yet:
* More than 2 workspaces (more than 1 underlay)
* Installing ROS 1 prerequisites for building the `ros1_bridge`
* RTI Connext - I don't yet have the special sauce wired in for leveraging the environment scripts

nuclearsandwich commented on doc/jobs/ci_jobs.rst:
```diff
@@ -0,0 +1,187 @@
+A *CI* job is used for continuous integration in batch.
+It builds the code and runs the tests to check for regressions.
+It operates on many source repositories as specified in one ore more ``.repos``
+files, and is triggered to produce a nightly "fat" archive, and can also be
```
I'd like to get away from using the word "fat" to describe these artifacts. I tried once before to settle on a replacement and then forgot. https://github.com/ros-infrastructure/rep/pull/184 avoids adding a descriptor to the archive altogether. Terms I've considered in the past are "comprehensive" and "omnibus" but I think a "workspace" archive is also reasonable.

nuclearsandwich commented on doc/jobs/ci_jobs.rst:
```diff
@@ -0,0 +1,187 @@
+The build process in detail
+---------------------------
+
+The actual build is performed within a Docker container in order to only make
```
Avoiding the suggestion API as the sentence is split across multiple lines but this doesn't quite scan for me.

I'll suggest
> The actual build is performed within a Docker container in order to ensure only declared dependencies are available.

nuclearsandwich commented on doc/jobs/ci_jobs.rst:
```diff
@@ -0,0 +1,187 @@
+non-underlay workspace. This means that any non-ROS packages present in that
+workspace may need their dependencies explicitly called out for inclusion in
+the foundation package list, and also means that a missing dependency may be
+occluded by another package in the workspace correctly taking the same
```
```suggestion
occluded by another package in the workspace correctly declaring the same
```

nuclearsandwich commented on doc/jobs/ci_jobs.rst:
```diff
@@ -0,0 +1,187 @@
+^^^^^^^^^^^^^^^^^^
+
+The following commands run the *CI* job for the *ament_cmake_ros* package
+from ROS *Crystal* for Ubuntu *Bionic* *amd64*:
```
What's the purpose of italicizing these words?

cottsay replied:
Totally just following the convention used in the other job docs. I think maybe the italics denote things that are configurable?

nuclearsandwich replied:
Thanks.

cottsay commented on scripts/ci/create_workspace.py:
```diff
@@ -57,8 +57,8 @@ def main(argv=sys.argv[1:]):
 
-    with Scope('SUBSECTION', 'vcs export --exact'):
-        export_repositories(args.workspace_root)
+#    with Scope('SUBSECTION', 'vcs export --exact'):
```
It was fixed in `vcstool` itself, as well as the package. We shouldn't see the problem again.

nuclearsandwich commented on ros_buildfarm/__init__.py:
```diff
@@ -17,4 +17,4 @@
 # - setup.py
 # - stdeb.cfg
-__version__ = '2.0.2-master'
+__version__ = '2.0.2-ci-builds'
```
Make sure this commit doesn't get merged.

nuclearsandwich commented on scripts/devel/build_and_test.py:
```diff
@@ -67,13 +67,14 @@ def main(argv=sys.argv[1:]):
                 args.workspace_root, 'test_results')
-            cmake_args = [
-                '-DBUILD_TESTING=1',
-                '-DCATKIN_ENABLE_TESTING=1', '-DCATKIN_SKIP_TESTING=0',
```
Has the diff on this file been deployed to standard devel jobs to check for regressions?

Was there a problem when specifying these arguments with colcon (for ROS 2)?

I'm slightly concerned about equating colcon use with ROS 2 even if that's the likely case today.

cottsay replied:
I believe the motivation is to reduce the number of CMake warnings because of unused command line definitions.

I don't think this is segregating the variables by ROS version - it just skips the `CATKIN_*` variables when we're using colcon.

dirk-thomas replied:
These arguments are still needed when building ROS 1 / catkin packages so they can't be removed. They are not specific to the build tool but to the build system used by the processed packages.

nuclearsandwich commented on scripts/devel/create_devel_task_generator.py:
```diff
@@ -99,45 +99,52 @@ def main(argv=sys.argv[1:]):
 
```
Nothing to critique. Just want to call attention to another change I think could have merited its own PR to cut down on the review burden and feature / noise ratio. Breaking this out would have also gotten it more testing on devel and PR jobs that aren't CI.

nuclearsandwich commented on setup.py:
```diff
@@ -44,6 +44,12 @@
     'long_description': long_description,
     'license': 'Apache 2.0',
+    'entry_points': {
+        'colcon_core.verb': [
```
I appreciate ros-buildfarm being in the verb name. I still find it confusing that the ros_buildfarm library installs colcon verbs but I may just have to get over that. I'm very much in favor of these verbs moving into the colcon org and being required by the jobs that need them.

cottsay replied:
Hopefully @dirk-thomas can lend his thoughts on if/where these verbs could land in upstream colcon. For the record, I agree, and I hope we can pull these out.

dirk-thomas replied:
I commented on the two verbs inline. I am not really convinced they are needed (and if one of them is that it should be in this repo).

cottsay replied:
The verbs have been dropped.

cottsay commented on ros_buildfarm/templates/devel/devel_task.Dockerfile.em:
```diff
@@ -70,11 +70,12 @@ RUN python3 -u /tmp/wrapper_scripts/apt.py update-install-clean -q -y ccache
+    install_lists=install_lists,
 ))@
 
+RUN update-ccache-symlinks
```
I hit this pretty early on, but it must boil down to the order in which things are installed. IIRC, I was consistently missing compiler symlinks when certain subsets of packages were built.

Add this specific re-linking resolved the issue, but we could drop it and see what breaks. If we get weird missing compiler issues, we know what the fix might look like...

cottsay commented on .travis.yml:
```diff
@@ -5,7 +5,7 @@ sudo: false
   global:
     - CONFIG_URL=https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/production/index.yaml
-    - CONFIG2_URL=https://raw.githubusercontent.com/ros2/ros_buildfarm_config/ros2/index.yaml
+    - CONFIG2_URL=https://raw.githubusercontent.com/ros2/ros_buildfarm_config/ci/index.yaml
```
Reminder to myself to change this after https://github.com/ros2/ros_buildfarm_config/pull/17 is merged.

dirk-thomas commented on ros_buildfarm/templates/snippet/reconfigure_jobs.groovy.em:
```diff
@@ -52,7 +52,7 @@ skipped_views = 0
 
 def view_dir = new File(view_config_dir)
-def views = view_dir.listFiles()
+def views = view_dir.listFiles() ?: []
```
This might still be worth doing.

Same below.

dirk-thomas commented on ros_buildfarm/workspace.py:
```diff
@@ -99,6 +99,8 @@ def call_build_tool(
         if os.path.exists(setup_file):
             cmd = '. %s && %s' % (setup_file, cmd)
+            if build_tool == 'colcon':
+                cmd = 'COLCON_CURRENT_PREFIX="%s" %s' % (parent_result_space, cmd)
```
This looks like a hack. Can you explain why that is done?

cottsay replied:
This would be necessary if the underlay workspace wasn't located in the exact same path in the container at which it was built. For most packages, setting `COLCON_CURRENT_PREFIX` is enough to let `setup.sh` know about the change.

I thought that this was sufficient until I tried building things on top of the rviz packages, which proved to have hard-coded paths making them non-relocatable. So I ended up re-structuring this job so that the underlay can sit at the same path, even though it shouldn't really be necessary.

So this can be removed now that we're limited to a single underlay and it sits in the same spot, but if any of that changes, this would be necessary since `setup.sh` can't detect its own location reliably, right?

dirk-thomas replied:
> This would be necessary if the underlay workspace wasn't located in the exact same path in the container at which it was built. For most packages, setting `COLCON_CURRENT_PREFIX` is enough to let `setup.sh` know about the change.

Ok, I understand now why you used it. So setting it to make the `.sh` relocatable sounds fine.

> I thought that this was sufficient until I tried building things on top of the rviz packages, which proved to have hard-coded paths making them non-relocatable. So I ended up re-structuring this job so that the underlay can sit at the same path, even though it shouldn't really be necessary.

Not sure if that is a good idea. The same problem would appear if users would download the underlay archive and unpack it elsewhere. Maybe it would actually be good to keep using a different location to actually catch that problem?

cottsay replied:
Agreed. I opened https://github.com/ros2/ros2/issues/606 to discuss the issue, and it sounds non-trivial. For now, I'd rather keep the workaround in place than render overlay jobs useless for a significant number of packages.

dirk-thomas replied:
:+1:

dirk-thomas commented on scripts/ci/build_task_generator.py:
```diff
@@ -0,0 +1,149 @@
+def main(argv=sys.argv[1:]):
+    parser = argparse.ArgumentParser(
+        description="Generate a 'Dockerfile' for the CI job")
+    parser.add_argument(
```
For a lot of the command line arguments the Python module `ros_buildfarm/argument` contains functions to add them. Please use them were possible and when declaring new arguments more than once move them into that module.

cottsay replied:
I surveyed all of the new scripts and could only find one spot where I could switch to one of the provided argument functions.

The biggest hold-up here is that the provided functions add os/arch/etc arguments as positional, where the pattern is to use the option-style in most places. Even though it means duplication, I think that deviating from the pattern used in other scripts with the same purpose is less desirable.

dirk-thomas replied:
> The biggest hold-up here is that the provided functions add os/arch/etc arguments as positional, where the pattern is to use the option-style in most places.

If the argument module only provides functions to add positional arguments (and they are used throughout the code - e.g. `add_argument_rosdistro_name()` 47 times) how can the option-style be the case used in most places? So I would argue that these new files should use the same style as the existing one which seem to be positional arguments for the required ones.

dirk-thomas commented on scripts/ci/create_workspace_task_generator.py:
```diff
@@ -0,0 +1,149 @@
+
+    # output hints about necessary volumes to mount
+    ros_buildfarm_basepath = os.path.normpath(
+        os.path.join(os.path.dirname(__file__), '..', '..'))
```
What is the purpose of this line?

cottsay replied:
Cruft. Removing. Good catch.

cottsay commented on doc/jobs/ci_jobs.rst:
```diff
@@ -0,0 +1,188 @@
+.. code:: sh
+
+  mkdir /tmp/ci_job
+  generate_ci_script.py https://raw.githubusercontent.com/ros2/ros_buildfarm_config/ros2/index.yaml crystal default ubuntu bionic amd64 --packages-select ament_cmake_ros --depth-before 100 > /tmp/ci_job/ci_job_crystal_ament_cmake_ros.sh
```
The default is to select only the packages passed to `--packages-select`. The `--depth-before` and `--depth-after` arguments expand the scope away from the specifically selected packages. So a sufficiently large value for `--depth-before` has similar behavior to `--packages-up-to`.

This minimizes the number of available arguments, while still supporting the "build packages from scratch" and the "build a subset of packages in an overlay" scenarios.

For a normal nightly build, this isn't necessary because `--packages-select` isn't specified, so all of the packages get built.

dirk-thomas replied:
Why not use `--packages-up-to ...` instead of `--depth-before 100 --packages-select ...`?

cottsay replied:
That's what I was saying - we can do that, but I'll need to plumb that value through from the script invocation and Jenkins parameters. Not a big deal, but non-zero work. The truth is that I'm just abusing the `scope` verb to obtain that behavior.

Now that I think about it, however, I think that using `--packages-up-to` would cause unexpected behavior when used in combination with `--depth-after`, because that mechanism expands on whatever packages were initially selected, which used to include only the named packages, but now includes all of the packages before them as well.

dirk-thomas replied:
> I think that using `--packages-up-to` would cause unexpected behavior when used in combination with `--depth-after`, because that mechanism expands on whatever packages were initially selected, which used to include only the named packages, but now includes all of the packages before them as well.

Maybe `--depth-after` should be changes into `--packages-above-depth PKG_NAME(s)` to avoid that problem?

cottsay replied:
Do you mean use `--packages-above PKG_NAME(s)` combined with a new argument `--packages-above-depth #`, where the default depth is infinite (current behavior)? I think that would work. If we drop `--depth-before` entirely as proposed in your other comment, we could then drop the `scope` verb.

dirk-thomas replied:
Correlating the `--packages-above` argument with the `--packages-above-depth` argument doesn't sounds appealing to me. E.g. you might want depth 1 above `foo` but also depth 2 above `bar`. So I would lean toward a new separate argument `--packages-above-depth N PKG_NAME(s)` (only usable once for now).

cottsay replied:
Another challenge we'll have here is that we want the OPPOSITE of this. The `scope` verb lets you select the packages you want, and everything else gets marked with `COLCON_IGNORE`. Do you think there is room for a package selection inversion mechanism, or do you see another way?

I feel that depth-after package selection is one of my favorite parts of this feature, and I want to get it right. If we can have that goodness in colcon itself, I'm all for it, I just want to get it right...

dirk-thomas replied:
> Another challenge we'll have here is that we want the OPPOSITE of this.

I am not sure I understand the use case you are referring to. Please elaborate.

> The `scope` verb lets you select the packages you want, and everything else gets marked with `COLCON_IGNORE`. Do you think there is room for a package selection inversion mechanism, or do you see another way?

You can achieve the same by running `colcon list` and remove `colcon list --packages-above-depth` from the result and then create ignore files for the remaining set of packages (if I understand the case correctly).

cottsay commented on doc/jobs/ci_jobs.rst:
```diff
@@ -0,0 +1,188 @@
+* Select the packages which should be built, tested, and installed
+
+  * Marks the packages explicitly listed as ignored
+  * If necessary, ignores any packages not selected or in scope
```
Yeah, this doesn't come across very clearly. My intent is to differentiate between the packages explicitly and unconditionally ignored, and the packages which are ignored because they aren't either selected (like `--packages-select`) or in scope (via `--depth-{before,after}`).

cottsay replied:
I clarified this in light of the changes to the scoping mechanism.

cottsay commented on doc/jobs/ci_jobs.rst:
```diff
@@ -0,0 +1,188 @@
+populate the workspace and enumerate prerequisites, one to perform the
+*build-and-install* task, and one to perform the *build-and-test* task.
+
+A typical job runs 4 Dockerfiles in all.
```
Sure, I'll remove it.

cottsay commented on doc/configuration_options.rst:
```diff
@@ -391,3 +391,40 @@ The following options are valid as keys in the ``_config`` dict under
+  default before any of the dependencies necessary to build the packages in the
+  workspace. Since not all packages in the workspace are necessarily ROS
+  packages, rosdep may be unable to detect and install the prerequisites for
+  those packages, so those prerequisite packages may need to be listed here.
```
My intent was to give us some flexibility by avoiding platform-specific concepts, but I'll just go ahead and make it implicit to conform with existing expectations.

cottsay commented on ros_buildfarm/colcon/verb/scope.py:
```diff
@@ -0,0 +1,101 @@
+        parser.add_argument(
+            'depth_before', type=int, metavar='NUM_BEFORE',
+            help='Number of forward-dependent packages upon which ' +
+                 'the targeted package(s) depends.')
```
I'd guess I don't really have a specific use case. Maybe I just misinterpreted something from our discussion on the desired package selection mechanisms for this feature.

If you see it as cruft, I can remove it, but I'll have to plumb a new mechanism to use use `--packages-up-to` instead, since some use cases involve "islands" of packages built on top of an underlay, and others involve building up from scratch.

dirk-thomas replied:
While I dislike the problems of `depth_after` I see when that is useful (build downstream packages to check if they still pass their tests after changes in a set of packages).

For `depth_before` I just don't see any use case. I don't see when "islands" of packages would be described by the depth before. If we don't have a case we want to use it for I would rather remove it.

dirk-thomas commented on ros_buildfarm/colcon/verb/ignore.py:
```diff
@@ -0,0 +1,47 @@
+            colcon_ignore = os.path.join(descriptor.path, IGNORE_MARKER)
+            if not os.path.isfile(colcon_ignore):
+                print('Creating %s' % (colcon_ignore))
+                open(colcon_ignore, 'a').close()
```
I would lean toward capturing the stdout from `colcon` and create the files in Python (simply because it works cross platform compared to `sed` / `xargs`).

> its way easier than trying to get data out of colcon programmatically.

If that is the case I would argue we can do whatever is necessary to make this easily accessible through an API.

cottsay commented on ros_buildfarm/ci_job.py:
```diff
@@ -0,0 +1,294 @@
+
+        'disabled': is_disabled,
+
+        # this should not be necessary
```
Good catch here. I can't find any reason to keep it, so I removed it.

cottsay commented:
Thanks @nuclearsandwich and @dirk-thomas for your feedback thus far. I've incorporated it and merged a series of spin-off PRs, and this PR is now ready to be reviewed once again. It's still pretty bug, but is much more focused. I would appreciate further feedback when you have some time.

cottsay commented on doc/jobs/ci_jobs.rst:
```diff
@@ -0,0 +1,194 @@
+.. code:: sh
+
+  mkdir /tmp/ci_job
+  generate_ci_script.py https://raw.githubusercontent.com/ros2/ros_buildfarm_config/ros2/index.yaml crystal default ubuntu bionic amd64 --packages-select ament_cmake_ros --depth-before 100 > /tmp/ci_job/ci_job_crystal_ament_cmake_ros.sh
```
These examples missed the update after the Colcon stuff got merged upstream. Will fix.

cottsay commented:
I moved all of the arguments used more than once to the common location. The `ci` task generation now differs from `devel` task generation in terms of argument format, but it matches the `release` task generation, which was the only job type that used positional arguments for task generations prior to this change.

I really like the idea of designing a system that can generate any number of CI jobs for any purpose, which can run on any schedule, and I think it is totally doable. Let's follow-up with a change that makes that happen - it should be possible to merge that work without affecting the "experience" presented in this PR.

dirk-thomas commented:
> Let's follow-up with a change that makes that happen - it should be possible to merge that work without affecting the "experience" presented in this PR.

Once CI is green (and this PR only adding new code but not really changing any existing code) I am fine to merge this (after reverting the custom branch names mentioned above - which will fail CI).

After that please follow up with a PR replacing the current fixed tuple of underlay / overlay jobs with the more flexible individual configuration.

cottsay commented:
I rebased and squashed the WIP commits, and removed the branch name and config override. CI will fail in this PR (and in master once merged) until ros2/ros_buildfarm_config#17 is merged.

nuclearsandwich commented:
Congrats on a long haul merge. :fireworks:
